### ISSUE TITLE
False positive protected-access in static method

### ISSUE THREAD

**Issue by mthuurne** [PyCQA/pylint#2825]
### Steps to reproduce

Run pylint on the following code:
```py
class Clazz:

    @staticmethod
    def create(value):
        instance = Clazz()
        instance.__value = value
        return instance

    def __init__(self):
        self.__value = None

    def __str__(self):
        return str(self.__value)

print(Clazz.create(123))
```

### Current behavior

The following warning is reported:
```
6:8: W0212: Access to a protected member __value of a client class (protected-access)
```
This is on the `instance.__value = value` statement.

### Expected behavior

I think no warning should be reported, since the static method is part of the same class that the protected member belongs to.

### pylint --version output

```
pylint 2.3.1
astroid 2.2.4
Python 3.5.6 (default, Feb  1 2019, 11:07:37) 
[GCC 7.3.1 20180323 [gcc-7-branch revision 258812]]
```

**Comment by PCManticore** [comment]
@mthuurne I believe a cleaner solution would be to provide a setter to your `Clazz` class instead of setting that member directly.

**Comment by mthuurne** [comment]
A public setter method means that anyone is allowed to change the value at any time, while setting the value in a factory method ensures that the value doesn't change after the object is created. So the two approaches are not equivalent.

Besides, even if you consider the code to be ugly, the warning message is still incorrect, since the code setting the value belongs to the same class as the value itself.

**Comment by vauhochzett** [comment]
Are @staticmethod methods not intended to access "protected" members of the class? I consider such methods explicitly part of the class. Same question goes for @classmethod methods, where this warning is emitted, too.

**Comment by Pierre-Sassoulas** [comment]
> A public setter method means that anyone is allowed to change the value at any time

You don't have to define the property setter, only defining the getter works.

> while setting the value in a factory method ensures that the value doesn't change after the object is created.

You can use an immutable class to make sure of this.

> Besides, even if you consider the code to be ugly, the warning message is still incorrect, since the code setting the value belongs to the same class as the value itself.

In
```
instance = Clazz()
instance.__value = value
return instance
```
``instance`` is not the ``Clazz`` instance (that would be 'self') so there's no false positive. The point of a value being private is so you can change the implementation in a single point and use the API everywhere else, clearly this is defeating the purpose.

If you want an immutable class sort out everything in the constructor and return the new instance.

```python
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class Clazz:
    value: Any


print(Clazz(123))
```

If you really insist on following the factory pattern ~~because (Redacted because we're friendly even with java enterprise developers)~~ you can still do:

```python
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class Clazz:
    value: Any

    @staticmethod
    def create(value):
        return Clazz(value)


print(Clazz.create(123))

```
